pkg/nl/list.go

```diff
@@ -1,149 +1,160 @@
+		if !(link.Type() == "vrf" && strings.HasPrefix(link.Attrs().Name, vrfPrefix)) {
+			continue
+		}
+		vrf := link.(*netlink.Vrf)
 
-			err := n.updateL3Indices(&info)
-			if err != nil {
-				return nil, err
-			}
+		info := VRFInformation{}
+		info.table = int(vrf.Table)
+		info.Name = link.Attrs().Name[3:]
+		info.vrfID = vrf.Attrs().Index
 
-			infos = append(infos, info)
+		err := n.updateL3Indices(&info)
+		if err != nil {
+			return nil, err
 		}
+
+		infos = append(infos, info)
 	}
 
 	return infos, nil
 }
 
-func (n *NetlinkManager) updateL3Indices(info *VRFInformation) error {
-	bridgeLink, err := netlink.LinkByName(BRIDGE_PREFIX + info.Name)
+func (*NetlinkManager) updateL3Indices(info *VRFInformation) error {
+	bridgeLink, err := netlink.LinkByName(bridgePrefix + info.Name)
 	if err != nil {
-		return err
+		return fmt.Errorf("error getting link by name: %w", err)
 	}
-	vxlanLink, err := netlink.LinkByName(VXLAN_PREFIX + info.Name)
+	vxlanLink, err := netlink.LinkByName(vxlanPrefix + info.Name)
 	if err != nil {
-		return err
+		return fmt.Errorf("error getting link by name: %w", err)
 	}
 	netlinkBridge := bridgeLink.(*netlink.Bridge)
 	netlinkVXLAN := vxlanLink.(*netlink.Vxlan)
 
-	info.bridgeId = netlinkBridge.Attrs().Index
+	info.bridgeID = netlinkBridge.Attrs().Index
 	info.VNI = netlinkVXLAN.VxlanId
 	return nil
 }
 
-func (n *NetlinkManager) updateL2Indices(info *Layer2Information, links []netlink.Link) error {
+func (*NetlinkManager) updateL2Indices(info *Layer2Information, links []netlink.Link) error {
 	for _, link := range links {
 		// Check if master of interface is bridge
 		if link.Attrs().MasterIndex != info.bridge.Attrs().Index {
 			continue
 		}
 
-		// If subinterface is VXLAN
-		if link.Type() == "vxlan" && strings.HasPrefix(link.Attrs().Name, VXLAN_PREFIX) {
-			info.vxlan = link.(*netlink.Vxlan)
-			info.VNI = info.vxlan.VxlanId
-		}
-
-		// If subinterface is VETH
-		if link.Type() == "veth" && strings.HasPrefix(link.Attrs().Name, VETH_L2_PREFIX) {
-			info.macvlanBridge = link.(*netlink.Veth)
-			peerIdx, err := netlink.VethPeerIndex(info.macvlanBridge)
-			if err != nil {
-				return err
-			}
-			peerInterface, err := netlink.LinkByIndex(peerIdx)
-			if err != nil {
-				return err
-			}
-			info.macvlanHost = peerInterface.(*netlink.Veth)
-			info.CreateMACVLANInterface = true
+		if err := updateLink(info, link); err != nil {
+			return err
 		}
 	}
 
 	// Read IP addresses
 	currentV4, err := netlink.AddrList(info.bridge, unix.AF_INET)
 	if err != nil {
-		return err
+		return fmt.Errorf("error listing link's IPv4 addresses: %w", err)
 	}
 	currentV6, err := netlink.AddrList(info.bridge, unix.AF_INET6)
 	if err != nil {
-		return err
+		return fmt.Errorf("error listing link's IPv6 addresses: %w", err)
 	}
-	for _, addr := range currentV4 {
+	for i, addr := range currentV4 {
 		if addr.Scope != unix.RT_SCOPE_UNIVERSE {
 			continue
 		}
-		info.AnycastGateways = append(info.AnycastGateways, &addr)
```

Comment:
@chdxD1 this was a bug that is getting fixed in this PR.
By simply referencing it this way, the same address is added multiple times: https://husni.dev/beware-of-implicit-memory-aliasing-in-go-foor-loop/